hazelcast/src/test/java/com/hazelcast/collection/impl/AbstractCollectionStatisticsTest.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.collection.impl;
+
+import com.hazelcast.collection.LocalCollectionStats;
+import com.hazelcast.test.HazelcastTestSupport;
+
+import static java.lang.String.format;
+import static org.junit.Assert.assertTrue;
+
+public abstract class AbstractCollectionStatisticsTest extends HazelcastTestSupport {
+
+    protected LocalCollectionStats localCollectionStats;
+
+    protected long previousAccessTime;
+    protected long previousUpdateTime;
+    protected long previousNumberOfHits;
+
+    protected void assertNewLastAccessTime() {
+        assertTrueEventually(() -> {
+            long lastAccessTime = localCollectionStats.getLastAccessTime();
+            assertTrue(format("Expected the lastAccessTime %d to be higher than the previousAccessTime %d (diff: %d ms)",
+                    lastAccessTime, previousAccessTime, lastAccessTime - previousAccessTime),
+                    lastAccessTime > previousAccessTime);
+            previousAccessTime = lastAccessTime;
+        });
+    }
+
+    protected void assertSameLastUpdateTime() {
+        assertTrueEventually(() -> {
+            long lastUpdateTime = localCollectionStats.getLastUpdateTime();
```

Comment:
I tried removing the `assertTrueEventually` block but the tests failed then. I don't fully understand why is it like that but I originally I did it this way because I went by the `QueueStatisticsTest` class. :)

Reply:
I'm not sure I fully understand your proposition. In the `ListStatisticsTest` class I check and save the `lastUpdateTime` after each operation. For example, before I test the `list.contains(o)` method, I save the previous `updateTime` (after testing the `list.get(0) method).

I suppose this check is not sufficient enough but I don't understand why. :(

(I am happy to make the changes you requested but wanted to ask for some clarification so I fully understand what's happening.)

Reply:
@aberkecz Sorry, I missed this question. The reason for the need for the `assertTrueEventually()` - let's take the test code snippet from my previous comment - is that by the time the `list.contains()` call returns, the operation's `afterRun()` might not be executed or completed (might run in parallel with the test's thread). You can check the reason here, note the order of sending the result and the call of `afterRun()`: https://github.com/hazelcast/hazelcast/blob/296aac5c8a3a4970851afc55c9e91622db75c131/hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/OperationRunnerImpl.java#L232-L242 Therefore, you can observe the change in the `lastAccessTime` only eventually. But since you don't expect any change for `lastUpdateTime`, the only way to tell for sure that no update happened to that field is if you order your assertion after the completion of the `afterRun()`. Apparently, there is nothing in the stats object that you can use for ordering the assertion with, so you need to perform two calls with the same arguments. The Hazelcast infrastructure takes care of that the two calls run sequentially, therefore, by the time the second call returns, the effect of the first call's `afterRun()` is visible, making it safe to assert on `lastUpdateTime`.

Reply:
Thanks @blazember for the detailed explanation! Now it's perfectly clear for me. The missing part for me was the fact that the `afterRun` method runs asynchronously. Let me fix the tests and get back to you ASAP.